web/source/text/processor.ts

```diff
@@ -30,49 +30,84 @@ namespace com.keyman.text {
     // Denotes whether or not KMW needs to 'swallow' the next keypress.
     swallowKeypress: boolean = false;
 
-
     /**
-     * Get the default key string. If keyName is U_xxxxxx, use that Unicode codepoint.
-     * Otherwise, lookup the  virtual key code (physical keyboard mapping)
+     * Get the default RuleBehavior for the specified key, attempting to mimic standard browser defaults 
+     * where and when appropriate.
      *
      * @param   {object}  Lkc  The pre-analyzed key event object
      * @param   {number}  keyShiftState
      * @param   {boolean} usingOSK
      * @return  {string}
      */
-    defaultKeyOutput(Lkc: KeyEvent, keyShiftState: number, usingOSK: boolean): string {
-      let outputTarget = Lkc.Ltarg;
-
+    defaultRuleBehavior(Lkc: KeyEvent, keyShiftState: number, usingOSK: boolean): RuleBehavior {
       let keyman = com.keyman.singleton;
 
+      let outputTarget = Lkc.Ltarg;
+      let preInput = Mock.from(outputTarget);
+      let ruleBehavior = new RuleBehavior();
+
       // check if exact match to SHIFT's code.  Only the 'default' and 'shift' layers should have default key outputs.
       if (keyShiftState == Codes.modifierCodes['SHIFT']) {
         keyShiftState = 1; // It's used as an index in some called methods.
       }
 
-      // If this was triggered by the OSK -or- if it was triggered by a 'synthetic' OutputTarget (TouchAlias, Mock)
-      // that lacks default key processing behavior.
+      let matched = false;
+      var char = '';
       if(usingOSK || outputTarget.isSynthetic) {
-        let result = DefaultOutput.applyCommand(Lkc, keyShiftState);
-
-        if(result) {
-          return result;
+        matched = true;  // All the conditions below result in matches until the final else, which restores the expected default
+                         // if no match occurs.
+
+        if(DefaultOutput.isCommand(Lkc)) {
+          // Note this in the rule behavior, return successfully.  We'll consider applying it later.
+          ruleBehavior.triggersDefaultCommand = true;
+
+          // We'd rather let the browser handle these keys, but we're using emulated keystrokes, forcing KMW
+          // to emulate default behavior here.
+        } else if(char = DefaultOutput.forSpecialEmulation(Lkc)) { 
+          switch(char) {
+            case '\b':
```

Comment:
If `forSpecialEmulation` output is used just here I'd prefer to see it return an enumerated set of values rather than a string/character.

Reply:
Could you please elaborate?  I'm not clear on what you mean here.

Reply:
It seems that `forSpecialEmulation` never returns a character that is passed on outside of KeymanWeb. In that case, it'd be better if it returned an enum type, e.g. `SpecialEmulation = { Backspace, Enter, ... }` or something like that -- because it restricts the return values to what can actually be returned.